utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,602 @@
+
+data VertexAttributeType
+  = Vec2
+  | Vec3
+  | Vec4
+  deriving (Show)
+
+vertexAttributeTypeFromText :: Text -> Maybe VertexAttributeType
+vertexAttributeTypeFromText = \case
+  "vec2" -> Just Vec2
+  "vec3" -> Just Vec3
+  "vec4" -> Just Vec4
+  _      -> Nothing
+
+vertexAttributeTypeToText :: VertexAttributeType -> Text
+vertexAttributeTypeToText = \case
+  Vec2 -> "vec2"
+  Vec3 -> "vec3"
+  Vec4 -> "vec4"
+
+instance FromJSON VertexAttributeType where
+  parseJSON = withTextMaybe "type" vertexAttributeTypeFromText
+
+instance ToJSON VertexAttributeType where
+  toJSON     = String . vertexAttributeTypeToText
+  toEncoding = toEncoding . vertexAttributeTypeToText
+
+convertVertexAttributeType :: VertexAttributeType -> (Word32, Format)
+convertVertexAttributeType = \case
+  Vec2 ->
+    (fromIntegral $ 2 * sizeOf (undefined :: Word32), FORMAT_R32G32_SFLOAT)
+  Vec3 ->
+    (fromIntegral $ 3 * sizeOf (undefined :: Word32), FORMAT_R32G32B32_SFLOAT)
+  Vec4 ->
+    ( fromIntegral $ 4 * sizeOf (undefined :: Word32)
+    , FORMAT_R32G32B32A32_SFLOAT
+    )
+
+data VertexAttribute = VertexAttribute
+  { binding  :: Word32
+  , location :: Word32
+  , size     :: Word32
+  , format   :: Format
+  }
+  deriving Show
+
+-- only support single binding for input vertex
+-- [SaschaWillems - Multiple Vulkan buffer binding points](https://gist.github.com/SaschaWillems/428d15ed4b5d71ead462bc63adffa93a)
+-- maybe can provide a binding map parameter to specify individual binding by name, like `foo [(1, ["inPos", "inColor"]) (2, ["texCoord"])]` speficies that `(binding = 1) inPos, (binding = 1) inColor, (binding = 2) texCoord`
+-- [island pipeline](https://github.com/tgfrerer/island/blob/76d0d38cba74181fa3774cef38aba4d96b6861dc/modules/le_backend_vk/le_pipeline.cpp#L21)
+makePipelineVertexInputStateCreateInfo
+  :: Vector Input -> Maybe (PipelineVertexInputStateCreateInfo '[])
+makePipelineVertexInputStateCreateInfo []     = Nothing
+makePipelineVertexInputStateCreateInfo inputs = Just zero
+  { vertexBindingDescriptions
+  , vertexAttributeDescriptions
+  }
+ where
+  vertexAttributes =
+    join . V.map makeVertexAttribute $ inputs :: Vector VertexAttribute
+  vertexAttributeDescriptions =
+    makeVertexInputAttributeDescriptions vertexAttributes :: Vector
+        VertexInputAttributeDescription
+  vertexBindingDescriptions =
+    makeVertexInputBindingDescriptions vertexAttributes :: Vector
+        VertexInputBindingDescription
+
+makeVertexInputBindingDescriptions
+  :: Vector VertexAttribute -> Vector VertexInputBindingDescription
+makeVertexInputBindingDescriptions =
+  V.fromList
+    . map makeVertexInputBindingDescription
+    . calculate
+    . groupBy ((==) `on` fst)
+    . sortOn fst
+    . extract
+ where
+  extract :: Vector VertexAttribute -> [("binding" ::: Int, "size" ::: Int)]
+  extract =
+    map
+        (   (,)
+        .   fromIntegral
+        .   (binding :: VertexAttribute -> Word32)
+        <*> fromIntegral
```

Comment:
Another place where `<*>`/`(.)` combination is really muddying the waters.

`map` and `toList` calls can be made in the parent code, so `extract` and `calculate` can be more simple per-element functions.

Reply:
fixed with
``` haskell
  where
    extract :: VertexAttribute -> ("binding" ::: Int, "size" ::: Int)
    extract = liftA2 (,) (fromIntegral . (binding :: VertexAttribute -> Word32)) (fromIntegral . size)
    calculate :: [ ("binding" ::: Int, "size" ::: Int) ]  -> ("binding" ::: Int, "stride" ::: Int) 
    calculate = liftA2 (,) (fst . head) (sum . (snd <$>))
```

the `liftA2 (,) foo1 foo2` here do the same thing as `(,) <$> foo1 <*> foo2` do, they can be both desugar to `(\input -> (foo1 input, foo2 input))`

Reply:
Yeah, I know. But this isn't obvious from the expression itself.
One has to consider its context and find out that it is a function application, which is the same as reader, which has an effect of "read the environment".

But the mental athletics doesn't stop here...

Unrolling `liftA2` / `(,)` we get:

```haskell
    calculate groups =
      ( (fst . head) groups
      ,  (sum . (snd <$>)) groups
      )
```

Unrolling a bit further...

```haskell
    calculate groups =
      ( fst $ head arg
      , sum $ map snd arg
      )
```

It's not that hard of course, but the reader has to do it every time... or give up and skip over this code.

And the argument itself is not obvious either. The `head` used here is guaranteed by call to `groupBy` before, but to get that information you have to pause here, take a look outside and then get back.

Reply:
fixed to non-pointfree style